net/tstun/wrap.go

```diff
@@ -325,26 +339,17 @@ func (t *Wrapper) IdleDuration() time.Duration {
 }
 
 func (t *Wrapper) Read(buf []byte, offset int) (int, error) {
-	var n int
-
-	wasInjectedPacket := false
-
-	select {
-	case <-t.closed:
+	if t.closed() {
 		return 0, io.EOF
-	case err := <-t.errors:
-		return 0, err
-	case pkt := <-t.outbound:
-		n = copy(buf[offset:], pkt)
-		// t.buffer has a fixed location in memory,
-		// so this is the easiest way to tell when it has been consumed.
-		// &pkt[0] can be used because empty packets do not reach t.outbound.
-		if &pkt[0] == &t.buffer[PacketStartOffset] {
-			t.bufferConsumed <- struct{}{}
-		} else {
-			// If the packet is not from t.buffer, then it is an injected packet.
-			wasInjectedPacket = true
-		}
+	}
+	res := <-t.outbound
```

Comment:
Ah, okay. poll and Read share one packet buffer, which circulates t.outbound -> Read -> t.bufferC -> poll -> (loop).

Mildly strange to me that the buffer in question is at t.buffer, where both routines can access it, and yet we're passing it around via this channel, rather than use the channel as a simple struct{} signalling chan. I suppose to signal close, we need it to be at least a bistate value, so chan bool where true means "you're free to use t.buffer again" and false means "closed". Am I missing something?

I think the way it's done here is also okay, but it feels like passing []byte around when it's actually just t.buffer being shared obscures the flow a little bit.